clippy_lints/src/unbound_return_lifetimes.rs

```diff
@@ -0,0 +1,141 @@
+    /// transmutes to the lifetime, the resulting lifetime may live longer
+    /// than was intended.
+    ///
+    /// **Known problems*: None.
+    ///
+    /// **Example:**
+    /// ```rust
+    /// struct WrappedStr(str);
+    ///
+    /// // Bad: unbound return lifetime causing unsoundnes
+    /// fn foo<'a>(x: impl AsRef<str> + 'a) -> &'a WrappedStr {
+    ///    let s = x.as_ref();
+    ///    unsafe { &*(s as *const str as *const WrappedStr) }
+    /// }
+    ///
+    /// // Good: bound return lifetime is sound
+    /// fn foo<'a>(x: &'a str) -> &'a WrappedStr {
+    ///   unsafe { &*(x as *const str as *const WrappedStr) }
+    /// }
+    /// ```
+    pub UNBOUND_RETURN_LIFETIMES,
+    correctness,
+    "unbound lifetimes in function return values"
+}
+
+declare_lint_pass!(UnboundReturnLifetimes => [UNBOUND_RETURN_LIFETIMES]);
+
+impl<'a, 'tcx> LateLintPass<'a, 'tcx> for UnboundReturnLifetimes {
+    fn check_item(&mut self, cx: &LateContext<'a, 'tcx>, item: &'tcx Item) {
+        if let ItemKind::Fn(ref sig, ref generics, _id) = item.kind {
+            check_fn_inner(cx, &sig.decl, generics, None);
+        }
+    }
+
+    fn check_impl_item(&mut self, cx: &LateContext<'a, 'tcx>, item: &'tcx ImplItem) {
+        if let ImplItemKind::Method(ref sig, _) = item.kind {
+            let parent_generics = impl_generics(cx, item.hir_id);
+            check_fn_inner(cx, &sig.decl, &item.generics, parent_generics);
+        }
+    }
+}
+
+fn check_fn_inner<'a, 'tcx>(
+    cx: &LateContext<'a, 'tcx>,
+    decl: &'tcx FnDecl,
+    generics: &'tcx Generics,
+    parent_generics: Option<(&'tcx Generics, Option<&'tcx Generics>, Option<&'tcx Generics>)>,
+) {
+    if let FunctionRetTy::Return(ref typ) = decl.output {
+        if let TyKind::Rptr(ref lifetime, _) = typ.kind {
+            if let LifetimeName::Param(_) = lifetime.name {
+                let target_lifetime = lifetime;
+
+                // check function generics
+                // the target lifetime parameter must appear on the left of some outlives relation
+                if let Some(param) = generics.get_named(target_lifetime.name.ident().name) {
+                    if param
+                        .bounds
+                        .iter()
+                        .any(|b| if let GenericBound::Outlives(_) = b { true } else { false })
+                    {
+                        return;
+                    }
+                }
+
+                // check parent generics
+                // the target lifetime parameter must appear on the left of some outlives relation
+                if let Some((ref parent_generics, _, _)) = parent_generics {
+                    if let Some(param) = parent_generics.get_named(target_lifetime.name.ident().name) {
+                        if param
+                            .bounds
+                            .iter()
+                            .any(|b| if let GenericBound::Outlives(_) = b { true } else { false })
```

Comment:
```suggestion
                            .any(|b| matches!(b, GenericBound::Outlives(_)))
```
Don't forge to append `#![feature(matches_macro)]` in crate root.